### ISSUE TITLE
getting UHBD potential from pdb

### ISSUE THREAD

**Issue by mwscandura** [Electrostatics/apbs-pdb2pqr#532]
multi-part issue here.
New to the software, trying to make an automated script for inputting pdb files and outputting UHBD grids.

general workflow for what I'm stuck on is
1. invoke pdb2pqr to create a subA.pqr file w/o water and w/ proper protonation, and create ABPS input file
`pdb2pqr --ff=amber --chain --drop-water --verbose --apbs-input --ph-calc-method=propka --with-ph=7.4 --propka-verbose $COREDIR/input/subA.pdb $WORKDIR/pdb2pqrA/subA.pqr`
outputs `subA.pqr` and `subA.in`

2. make APBS of PQR file w/ electrostatic potential grids using APBS input file
`apbs subA.in`
outputs `subA.pqr.dx`

the issue is that i need the output to be in UHBD format, not OPENDX

I've scoured the documentation and see that UHBD is a supported output format (section 8.1.3) but don't understand how to do this. I understand that in the ABPS input file ELEC section, the line `write pot dx`  can be changed to `write pot uhbd`, but doing this manually i get this:
`Vgrid_writeUHBD: can't write UHBD mesh with non-uniform spacing`

SO, short version of the question is how do i generate a UHBD output starting from a PDB?
Longer version is:
- is there a way to automate `subA.in` automatically be written to have line, `write pot uhb`d (for example, something during the PDB2PQR step?) or must it be changed manually every time?
- what is the meaning of the` can't write UHBD mesh with non-uniform spacing` ?
- might the dx2uhbd tool be useful here? As a novice I'm not sure how to alter it for my needs (or even what kind of script it is - looks maybe like java?)

**Comment by kozlac** [comment]
Hello @mwscandura, if you look at the contents of the `subA.in` file you should see the cglen and fglen keywords. In order for APBS to output the potential map in UHBD format these (x,y,z) values should be the same in each direction (Hence the non-uniform error). The cglen dimensions should be large enough to encompass the entire molecule while the fglen ones should cover the area of interest. Please see [this link](https://apbs-pdb2pqr.readthedocs.io/en/latest/apbs/input/elec/mg-auto.html#mgauto) for a description of the keywords.

There's no PDB2PQR step to automatically write the in file to write in UHBD format. You could write a python script to take the max in each of the cglen and fglen, make them uniform, and change the write line. Please let me know if you need help with that.

**Comment by mwscandura** [comment]
I see. Any help or guidance on writing the python script would be excellent, I'm not familiar with the langauge

**Comment by kozlac** [comment]
Hello @mwscandura, the following script will take all the `.in` files in a folder an change the `cglen` and  `fglen` to the corresponding max's and change the `write` format to uhbd as long as you have python 3.5 or higher installed. For example in an Ubuntu system just put the script in the same folder as all the `.in`  files and run: `python3  convert2uhbd.py`. 

[convert2uhbd.zip](https://github.com/Electrostatics/apbs-pdb2pqr/files/3912746/convert2uhbd.zip)



Let me know if you have any troubles. Thanks!!

**Comment by mwscandura** [comment]
Thanks again for this, Juan. I’ve only just had a chance to take a look and try to get it working with what I’m doing. It didn’t seem to go through in my first attempt and I got a few error messages back that maybe you can clear up:

line 6: import: command not found
line 8: syntax error near unexpected token `('
line 8: `def getMax(l):'

I’m running python 3.7.3

***  ***  ***  ***  ***  ***  ***  ***
Matthew Scandura
Postbaccalaureate Fellow, IRTA
Laboratory of Immune System Biology (LISB)
National Instittue of Allegy & Infectious Diseases (NIAID)
National Institutes of Health (NIH), HHS

**Comment by kozlac** [comment]
Marking it as resolved.

**Comment by mwscandura** [comment]
Hello Juan,

I’m unsure if you’ll still see replies in response to this. We corresponded over Github a few months ago regarding an issue with outputting UHBD formatted files in APBS.
Thanks again for the script. I made a small modification (turns out the dime also have to all be equal) and I’ve integrated it into my pipeline. I’m a researcher at the NIH this pipeline is a major part of my project, which I’ll be presenting to my lab group about next week. I’d like to include you in the acknowledgments and wanted to ask how to most appropriately do so.

Thanks,
Matthew


***  ***  ***  ***  ***  ***  ***  ***
Matthew Scandura
Postbaccalaureate Fellow, IRTA
Laboratory of Immune System Biology (LISB)
National Institute of Allergy & Infectious Diseases (NIAID)
National Institutes of Health (NIH), HHS

**Comment by kozlac** [comment]
Hello @mwscandura. I'm really glad I was able to help.  I think Juan Brandi, Computer Scientist with the APBS development team would do. Thanks!!

**Comment by mwscandura** [comment]
Will do!


***  ***  ***  ***  ***  ***  ***  ***
Matthew Scandura
Postbaccalaureate Fellow, IRTA
Laboratory of Immune System Biology (LISB)
National Institute of Allergy & Infectious Diseases (NIAID)
National Institutes of Health (NIH), HHS
